### ISSUE TITLE
Please support marking different semantic regions in diff lines

### ISSUE THREAD

**Issue by joshtriplett** [libgit2/libgit2#3874]
The `git_diff_line` struct represents one or more lines of a diff.  However, a single line of a diff may contain multiple semantic regions, which a frontend may want to format or color differently.  For instance, a line with `origin == GIT_DIFF_LINE_HUNK_HDR` may include both a hunk header (between the @@ markers) and a function (after the second @@ marker); git-diff colors those with two different styles.  Similarly, a line with `origin == GIT_DIFF_LINE_ADDITION` may include regions of whitespace errors.  And a file header with `origin == GIT_DIFF_LINE_FILE_HDR` can actually include multiple lines: the `diff --git` line, the `index` line, lines about file modes, and the `---` and `+++` lines.

Currently, frontends that want to handle those regions differently have to carefully parse the text of the line.

Please consider extending git_diff_line to allow iterating over a list of regions within the line.  Each region would include a pointer within the content, a length, and an enum for the type.  (The concatenation of all the regions should always match the full content of the line.)  This would allow frontends to walk the regions and emit each one (potentially with formatting).

**Comment by ethomson** [comment]
How does `git-diff` do this?  Does xdiff emit this information, or is it parsing those lines itself?

**Comment by joshtriplett** [comment]
@ethomson As far as I can tell, git-diff only handles this for a single case: the func appearing after the second `@@` in a hunk header.  For that case, git-diff has some fragile parsing of the line.  (As far as I can tell, that parsing would break for a merge commit with `@@@ ... @@@ ...`, for instance.)

**Comment by ethomson** [comment]
Right, that's what I would have expected.  Most consumers of this are parsing the `git-diff` output themselves.
